[HotA] Add aviation

mrhaandi opened the pull request:
Fixes #4051. Too big to be ready for 1.7.

Why it is difficult:
- destination tile and start mode does not specify layer uniquely anymore (for boats: if you sail and click on a coast tile, then you want to disembark)
- if you click on a land tile while in an airship both landing and flying there is a legal possibility (current `MoveHero` packet cannot capture the difference)
- aviation has different rules from fly (similar how sailing `SAIL` is different from water walking `WATER`)

**TASKS for Airship**
- [x] add EPathfindingLayer `AVIATE` (different from `AIR`)
- [x] add transitions: `LAND` <-> `AVIATE` (hover over accessible land and water tiles) <-> `FLY` (cannot stop over obstacle)
- [x] add layer specification to `MoveHero` destination handling
- [x] add path finder `AVIATE` rules
- [x] add node storage `AVIATE` specifications
- [x] add adventure map aviation interactions
- [x] add EAdventureState `DISEMBARKING`
- [x] add disembark shortcut Shift+D for `ADVENTURE_DISEMBARK`
- [x] add HotA cursor icons for airship travel
- [x] while in airship hero can get attacked
- [x] correct move point generation / consumption
- [x] player testing
- [x] ai testing (no crashes, no ai heroes caught in airships)
- [ ] add [airShip.zip](https://github.com/user-attachments/files/24822393/airShip.zip) with configuration and icons to HotA mod

**TASKS for Airship Yard**
- [x] can buy an airship from Airship Yard on land and coast
- [x] set proper cost for airship

**NOTES**
- Since the aviation cursor icons are part of main code, they are included in `Mods\vcmi\Content\Sprites\airshipCursor` (not HotA).

To not delay the core features, I will not include the following features.
These can be added later separately.
- configurable (air)ship cost
- configurable (air)ship movement points
- support free airship boarding bonus
- remove tearing in airship preview
- ai taking airship travel into account
- add dedicated disembark adventure map button

mrhaandi commented:
> Does it mean the functions of airship might be finished before 1.7 release, but airship yard might be put on hold?

Unfortunately, no. The changes required for aviation are quite profound and touch code of basic travel mechanics, even if no airship is present. It would be unwise to release it too hastily into 1.7. This holds true, even if I'd manage to complete the code before 1.7 release.

GeorgeK1ng commented:
So almost ready right? Then mod support and AI remains?

mrhaandi commented:
Rebased and retargeted to `beta` as per request of @GeorgeK1ng

mrhaandi commented:
> Great, can you also please run CI to get new artifacts for testing please?

I believe, I do not have permission to manually run VCMI CI on the PR.
I manually ran CI on the rebased branch: https://github.com/mrhaandi/vcmi/actions/runs/21619796664

GeorgeK1ng commented:
This one for 1.7.4 would be great, seems there are some conflicts which needs to be resolved.

IvanSavenko commented:
Will take a look. Need to review it first, some changes are clearly still necessary

GeorgeK1ng commented:
@by003 Here is mentioned installer with AirShip support.

> > Great, can you also please run CI to get new artifacts for testing please?
> 
> I believe, I do not have permission to manually run VCMI CI on the PR. I manually ran CI on the rebased branch: https://github.com/mrhaandi/vcmi/actions/runs/21619796664

mrhaandi committed:
add aviation

- add EPathfindingLayer `AVIATE` (different from `AIR`, can hover over free land and water)
- add transitions: `LAND` <-> `AVIATE` <-> `AIR` (cannot stop over obstacle)
- add layer specification to `MoveHero` destination handling
- add path finder `AVIATE` rules
- add node storage `AVIATE` specifications
- add adventure map aviation interactions
- add EAdventureState `DISEMBARKING`
- add disembark shortcut Shift+D for `ADVENTURE_DISEMBARK`
- add cursor icons for aviation under CC-BY-SA license
- correct move point generation / consumption
- can buy an airship from Airship Yard on land and coast at fixed cost
- add localization

mrhaandi commented:
I rebased the PR on top of the current beta to resolve conflicts. The conflicts were due to upstream relocation of translation files.

GeorgeK1ng commented:
Great! @by003 You can download installer from this PR now directly.

by003 commented:
Thanks for sharing! I will test it few days.
<img width="867" height="649" alt="image" src="https://github.com/user-attachments/assets/97c0cba6-6aa7-4c85-8ac8-f25780ac3b74" />
Which one can be used on PC?

GeorgeK1ng commented:
Best to check and download your wanted platform from Action Summary, we have beautiful report for all platforms :)
<img width="704" height="257" alt="image" src="https://github.com/user-attachments/assets/a49137ea-795d-40f1-a14c-381da6535784" />

by003 commented:
<img width="1128" height="976" alt="image" src="https://github.com/user-attachments/assets/8181f816-0c24-4f53-bfcc-1fa7dfd35712" />
My system is x64.

by003 commented:
> @by003 please attach crash dump, it's saved near the exe

That's fine, I just want to find the PR which can be run in my PC

kambala-decapitator commented:
sorry, I thought that was screenshot of a crash log :)

@GeorgeK1ng has already shown you what to download on his screenshot

rjulius23 commented:
When is it coming ? I mean airships are kind of broken in the current main.  I mean in 1.7.3 it is broken, you cannot disembark the Airship or was that intended ?

IvanSavenko commented on lib/networkPacks/PacksForServer.h:
```diff
@@ -49,13 +49,13 @@ struct DLL_LINKAGE DismissHero : public CPackForServer
 	{
 	}
-	std::vector<int3> path;
+	std::vector<std::pair<int3, EPathfindingLayer>> path;
```
I think we can keep the code simple if entire pack (and entire move) would he made within a single layer:
```cpp
std::vector<int3> path;
EPathfindingLayer layer;
```
...and for long moves - send chain of packs, with different layers

Client already makes such batching, but mostly for faster movement in mp

mrhaandi replied:
Will try.

mrhaandi replied:
Now, I adapted the `MoveHero` pack to have a uniform layer argument, as per your suggestion.
In `HeroMovementController::sendMovementRequest` batches are now separated also on layer switch.

IvanSavenko commented on client/PlayerLocalState.h:
```diff
@@ -77,7 +77,7 @@ class PlayerLocalState
 
 	void setPath(const CGHeroInstance * h, const CGPath & path);
-	bool setPath(const CGHeroInstance * h, const int3 & destination);
+	bool setPath(const CGHeroInstance * h, const int3 & destination, const EPathfindingLayer & layer = EPathfindingLayer::AUTO);
```
Any way to avoid this default parameter and set layer explicitly in every call? In my experience various autoselection often leads to obscure bugs at some point later

mrhaandi replied:
Now the default parameter value is removed and must be set explicitly.

IvanSavenko commented on client/adventureMap/AdventureMapInterface.cpp:
```diff
@@ -881,6 +936,45 @@ void AdventureMapInterface::openWorldView(const std::vector<ObjectPosInfo>& obje
 	widget->getMapView()->onViewSpellActivated(11, objectPositions, showTerrain);
 }
 
+void AdventureMapInterface::enterDisembarkMode()
+{
+	setState(EAdventureState::DISEMBARKING);
+}
```
Does hota also has such "disembark mode" logic or they use some other approach?

mrhaandi replied:
Yes, HotA has a dedicated disembark mode (to a neighboring tile). Unlike boats which cannot go on land, it is necessary to distinguish movement from disembarking.

IvanSavenko commented on lib/mapObjects/CGHeroInstance.cpp:
```diff
@@ -213,6 +213,7 @@ int CGHeroInstance::movementPointsLimit() const
 
 int CGHeroInstance::movementPointsLimit(bool onLand) const
 {
+	//FIXME wrong when hero is in airship (not on land, not in water)
```
Probably idea of method is wrong.
Instead it should:
- accept no parameters 
- return land limit when hero is not in boat
- return limit defined by boat if hero is in boat

mrhaandi replied:
Adapted as per your suggestion, fixing the `FIXME`. The `bool onLand` argument is removed.
There are cases in which a layer argument is explicitly required e.g. in the ai method
`int ChainActor::maxMovePoints(CGPathNode::ELayer layer)`
In such cases I use `return hero->getTurnInfo(0)->getMaxMovePoints(layer);`

IvanSavenko commented on lib/mapObjects/CGHeroInstance.cpp:
```diff
@@ -242,12 +243,19 @@ std::unique_ptr<TurnInfo> CGHeroInstance::getTurnInfo(int days) const
+	if (layer == EPathfindingLayer::LAND)
 		return ti->getMovePointsLimitLand();
-	else
+	else if (layer == EPathfindingLayer::WATER || layer == EPathfindingLayer::SAIL)
```
I think we should only check for sail here

Movement limit for water layer is not defined, and accessing it is an error

mrhaandi replied:
Fixed.

IvanSavenko commented on lib/mapObjects/CGHeroInstance.cpp:
```diff
@@ -1369,6 +1377,9 @@ int CGHeroInstance::movementPointsAfterEmbark(int MPsBefore, int basicCost, bool
 	
 	auto boatLayer = inBoat() ? getBoat()->layer : EPathfindingLayer::SAIL;
 
+	if(boatLayer == EPathfindingLayer::AVIATE)
+		return 0; // boarding an airship takes all MPs
+
```
Probably free ship boarding bonus should have pathfinding layer as a subtype? To allow airship version of admiral hat?

mrhaandi replied:
Generally, yes it is a good idea. Due to my currently limited capacity, I would just add a comment regarding this, and leave as currently is.

IvanSavenko commented on lib/mapObjects/IObjectInterface.cpp:
```diff
@@ -92,7 +92,7 @@ int3 IBoatGenerator::bestLocation() const
 			continue; // tile not visible / outside the map
 
-		if(!tile->isWater())
+		if((layer == EPathfindingLayer::SAIL || layer == EPathfindingLayer::AUTO) && !tile->isWater())
```
In which cases layer may be AUTO? IMO this is illegal - layer should be clearly specified by caller

mrhaandi replied:
Fixed.

IvanSavenko commented on lib/mapObjects/IObjectInterface.h:
```diff
@@ -103,7 +103,7 @@ class DLL_LINKAGE IBoatGenerator
 	virtual BoatId getBoatType() const = 0; //0 - evil (if a ship can be evil...?), 1 - good, 2 - neutral
 	virtual void getOutOffsets(std::vector<int3> & offsets) const = 0; //offsets to obj pos when we boat can be placed
-	int3 bestLocation() const; //returns location when the boat should be placed
+	virtual int3 bestLocation(const EPathfindingLayer & layer = EPathfindingLayer::AUTO) const; //returns location when the boat should be placed
```
Please avoid default parameters, especially magical autoselection

mrhaandi replied:
I removed the parameter completely. The layer is a property of the ship creator now.

IvanSavenko commented on lib/pathfinder/CPathfinder.cpp:
```diff
@@ -356,8 +361,13 @@ bool CPathfinder::isLayerTransitionPossible() const
 
 		break;
 
+	case ELayer::AVIATE:
+		if((destLayer == ELayer::LAND && !destination.tile->isWater()) || destLayer == ELayer::AIR)
+			return true;
```
We can go from airship straight into fly mode? Is this hota logic?

mrhaandi replied:
Yes. The difficult part about the airship is that it may hover over land (layer `AVIATE`) but cannot hover e.g. over mountains. For this, the layer `FLY` is used (while in airship) and it has the same mechanics as the fly spell for going over obstacles in one go to an accessible tile.

IvanSavenko commented on lib/constants/EntityIdentifiers.h:
```diff
@@ -733,7 +733,7 @@ class DLL_LINKAGE EPathfindingLayerBase : public IdentifierBase
 	enum Type : int32_t
 	{
-		LAND = 0, SAIL = 1, WATER, AIR, NUM_LAYERS, WRONG, AUTO
+		LAND = 0, SAIL = 1, WATER, AVIATE, AIR, NUM_LAYERS, WRONG, AUTO
```
Honestly, looking through changes I think we should nuke AUTO and always specify layer clearly, to avoid unexpected automagic changes of layer

mrhaandi replied:
Will try.

mrhaandi replied:
After a preliminary look, it looks like there are distinct legitimate uses of `AUTO` both for don't know (clicking on a tile in the interface code) and don't care (town portal) cases.
Generally, it is a good idea to have a consistent, precise, and minimal use of `AUTO` as default. However, it is somewhat orthogonal to aviation in this PR.
I will consider this for a subject of a later PR, but would not compound it here.

mrhaandi commented:
@IvanSavenko I believe all your review comments are addressed now, in one way or another.

IvanSavenko commented on lib/callback/CCallback.h:
```diff
@@ -40,8 +40,8 @@ class DLL_LINKAGE CCallback : public CPlayerSpecificInfoCallback, public CBattle
-	void moveHero(const CGHeroInstance *h, const std::vector<int3> & path, bool transit) override;
-	void moveHero(const CGHeroInstance *h, const int3 & destination, bool transit) override;
+	void moveHero(const CGHeroInstance *h, const std::vector<int3> & path, const EPathfindingLayer & layer, bool transit) override;
+	void moveHero(const CGHeroInstance *h, const int3 & destination, bool transit, const EPathfindingLayer & layer = EPathfindingLayer::AUTO) override;
```
Please make order of parameters consistent - right now one override has `layer, transit`, while another has `transit, layer`

mrhaandi replied:
Done.

IvanSavenko commented on lib/mapObjects/CGHeroInstance.cpp:
```diff
@@ -1099,6 +1097,11 @@ BoatId CGHeroInstance::getBoatType() const
 	return BoatId(LIBRARY->townh->getById(getHeroClass()->faction)->getBoatType());
 }
 
+EPathfindingLayer CGHeroInstance::getBoatLayer() const
+{
+	return EPathfindingLayer::SAIL;
+}
+
```
In theory, this would prevent creation of "summon airship" kind of spells. But most likely was a problem before, just in a different location in code, and out of PR scope.

mrhaandi replied:
Yes.

IvanSavenko commented on client/adventureMap/AdventureMapInterface.cpp:
```diff
@@ -628,6 +646,24 @@ void AdventureMapInterface::onTileHovered(const int3 &targetPosition)
 
+	if(getState() == EAdventureState::DISEMBARKING)
+	{
+		Cursor::Map cursor_index = Cursor::Map::POINTER;
```
camelCase for local variables
```suggestion
		Cursor::Map cursorIndex = Cursor::Map::POINTER;
```

mrhaandi replied:
Done.

IvanSavenko reviewed (approved):
Generally looks OK. Potential remaining TODO's. Would be great to solve them in this PR, but if your are planning to fix them in next PR - that should be OK too.

- Hardcoded cost of Airship
- Lack of free ship boarding bonus support for Airship
- Movement point limit for both ships and airships should be defined by ship, not hardcoded

mrhaandi commented:
> Would be great to solve them in this PR, but if your are planning to fix them in next PR - that should be OK too.

Before getting too deep into generalization, I would like to get the core functionality running as well as have feedback from users.
To keep track for a future PRs, I added your points to the description.

IvanSavenko commented:
In that case - OK for merge into 1.8 / develop. But too many changes for 1.7.4 / beta.

If you rebase it into develop & retarget PR to develop branch - I can merge it immediately, if not - will merge it myself a bit later.

mrhaandi commented:
> If you rebase it into develop & retarget PR to develop branch - I can merge it immediately, if not - will merge it myself a bit later.

I squashed, rebased on develop, and retargeted.
